### DISCUSSION TITLE
sqlalchemy.orm.synonym doesn't seem to work with functions bulk_insert_mappings and execute

### DISCUSSION THREAD

**Discussion by flyly0755** [sqlalchemy/sqlalchemy#10439] [Usage Questions] [answered]
There is such a situation, data of different sources need to insert to one table, but the data column is a little different, for example:
Data from source a has column fed, but data from source b has column Fed, 
or data from source a has column fed1, but data from source b has column fed2.
now i want to create a table with column fed to process all kind of columns such as Fed, fed1, fed2.
Create ORM as below:
```python
from sqlalchemy import create_engine
from sqlalchemy import Column
from clickhouse_sqlalchemy.ext.declarative import declarative_base
from clickhouse_sqlalchemy import types, engines
from sqlalchemy.orm import sessionmaker, synonym

ChBase = declarative_base()


class CaseTest(ChBase):
    abc = Column(types.Int8, primary_key=True)
    fed = Column(types.String)
    Fed = synonym('fed')
    fed1 = synonym('fed')
    fed2 = synonym('fed')
    __tablename__ = 'caseTest2'
    __table_args__ = (
        engines.MergeTree(order_by='abc', primary_key='abc'),
        {'comment': 'synonym test'}
    )

ckuser = "user"
ckpwd = "password"
ckip = "host"
ckport = "8123"
ckdbname = 'adbname'

uri = f"clickhouse://{ckuser}:{ckpwd}@" \
      f"{ckip}:{ckport}/{ckdbname}"

engine = create_engine(uri, echo=False)
DBsession = sessionmaker(bind=engine)
session = DBsession()
testdata = [{'abc': 1, 'fed': 'x'}, {'abc': 2, 'Fed': 'y'}, {'abc': 3, 'fed1': 'z'}]
```
try with session.add(CaseTest(Fed='z', abc=3)) which can be insert successfully and result is abc:3, fed:z.
but with session.execute(insert(CaseTest), testdata) and session.bulk_insert_mappings(CaseTest, testdata) both not work
session.execute raise error:
```error
sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'fed', in parameter group 1
[SQL: INSERT INTO "caseTest2" (abc, fed) VALUES (%(abc)s, %(fed)s)]
[parameters: [{'abc': 1, 'fed': 'x'}, {'abc': 2, 'Fed': 'y'}]]
(Background on this error at: http://sqlalche.me/e/14/cd3x)
```
session.bulk_insert_mappings(CaseTest, testdata) run successfully but actual result is Fed is ignored, and fed value in database table is empty with raw data  {'abc': 2, 'Fed': 'y'}
Is there any method can achieve the result as function session.add without any extra data transformation, such as
transfer all fed releated columns which is not fed to fed(Fed-->fed, fed1-->fed), this is cumbersome.
Some extra config in ORM or parameter in functions execute and bulk_insert_mappings can solve it?

**Comment by CaselIT** [comment]
Hi,

Not at the moment, my suggestion is for now to create a list of instances and then add them to a session using `add_all`.

@zzzeek do you think it's feasible to support this use case when using insert?

**Reply by flyly0755** [reply]
yeah, add_all can be a backup solution, lines of code will be short.
```python
testdata = [{'abc': 5, 'fed': 'o'}, {'abc': 6, 'Fed': 'p'}]
ins_list = []
for item in testdata:
    ins_list.append(CaseTest(**item))
session.add_all(ins_list)
```
but I think with big data volumn this solution seems memory inefficient and take extra time used in instance loop creation.
If synonym feature can support the scenario I met, this will be a big convenience.
Current synonym function only supports session.add seems big limitations with it.

**Reply by zzzeek** [reply]
bulk ORM insert should probably support naming attributes as synonyms, I wonder if this worked in 1.4 and this is a regression?  seems strange synonyms never worked for the old method from 1.x.

**Reply by CaselIT** [reply]
will you check it mike or are you leaving it as a question?

**Reply by zzzeek** [reply]
I...was leaving it open for someone to triage at least :)   that is, is this a regression from 1.4 or not.    we would implement for 2.0.x.   hopefully there's nothing too tricky about it, not sure.

**Reply by CaselIT** [reply, marked as answer]
It appeart that in 1.4 this was not working.
Running `session.execute(insert(CaseTest), testdata)`
Version 2 ignores the synonym and splits the insert into two:
```
2023-10-09 20:50:15,264 INFO sqlalchemy.engine.Engine
CREATE TABLE "caseTest2" (
        abc INTEGER NOT NULL,
        fed VARCHAR,
        PRIMARY KEY (abc)
)


2023-10-09 20:50:15,264 INFO sqlalchemy.engine.Engine [no key 0.00013s] ()
2023-10-09 20:50:15,265 INFO sqlalchemy.engine.Engine INSERT INTO "caseTest2" (abc, fed) VALUES (?, ?)
2023-10-09 20:50:15,266 INFO sqlalchemy.engine.Engine [no key 0.00013s] (1, 'x')
2023-10-09 20:50:15,266 INFO sqlalchemy.engine.Engine INSERT INTO "caseTest2" (abc) VALUES (?)
2023-10-09 20:50:15,266 INFO sqlalchemy.engine.Engine [no key 0.00010s] [(2,), (3,)]
```
version 1.4 instead
```
sqlalchemy.exc.InvalidRequestError: A value is required for bind parameter 'fed', in parameter group 1 (Background on this error at: https://sqlalche.me/e/14/cd3x)
```

running `session.bulk_insert_mappings(CaseTest, testdata)` instead has the same behaviour on both 1.4 and 2.0:
```
CREATE TABLE "caseTest2" (
        abc INTEGER NOT NULL,
        fed VARCHAR,
        PRIMARY KEY (abc)
)


2023-10-09 20:52:50,929 INFO sqlalchemy.engine.Engine [no key 0.00014s] ()
2023-10-09 20:52:50,930 INFO sqlalchemy.engine.Engine INSERT INTO "caseTest2" (abc, fed) VALUES (?, ?)
2023-10-09 20:52:50,930 INFO sqlalchemy.engine.Engine [generated in 0.00014s] (1, 'x')
2023-10-09 20:52:50,930 INFO sqlalchemy.engine.Engine INSERT INTO "caseTest2" (abc) VALUES (?)
2023-10-09 20:52:50,930 INFO sqlalchemy.engine.Engine [generated in 0.00010s] ((2,), (3,))
```

so not a regression. Will open an issue as use case
